File: python/medium/2523.closest_prime_numbers_in_range/closest_prime_pair_in_range.py
from typing import List

class Solution:
    def closestPrimes(self, left: int, right: int) -> List[int]:
        all_primes = [True] * (right+1)
        i = 2
        while i*i <= right:
            if all_primes[i]:
                for j in range(i*i, right+1, i):
                    all_primes[j] = False
            i+=1
        
        primes = []
        for i in range(2, len(all_primes)):
            if all_primes[i] and i >= left:
                primes.append(i)
            
        answer = [-1, -1]
        min_diff = float('inf')
        for i in range(1, len(primes)):
            diff = primes[i] - primes[i-1]
            if diff < min_diff:
                min_diff = diff
                answer[0] = primes[i-1]
                answer[1] = primes[i]
        
        return answer

File: c++/medium/2523.closest_prime_numbers_in_range/closest_prime_pair_in_range.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> closestPrimes(int left, int right) {
        vector<bool> allPrimes(right+1, true);

        for(int i = 2; i*i <= right; i++) {
            if(allPrimes[i]) {
                for(int j = i*i; j <= right; j+=i) {
                    allPrimes[j] = false;
                }
            }
        }

        vector<int> primes;
        for(int i = 2; i < allPrimes.size(); i++) {
            if(allPrimes[i] && i >= left) {
                primes.push_back(i);   
            }
        }

        vector<int> answer = {-1, -1};
        int minDiff = INT_MAX;
        for(int i = 1; i < primes.size(); i++) {
            int diff = primes[i] - primes[i-1];
            if(diff < minDiff) {
                minDiff = diff;
                answer[0] = primes[i-1];
                answer[1] = primes[i];
            }
        }

        return answer;
    }
};
